optimum/exporters/executorch/integrations.py

```diff
@@ -24,15 +24,274 @@
+        """
+        Export the multimodal model into separate ExecuTorch programs.
+
+        Returns:
+            Dict[str, ExportedProgram]: Dictionary containing exported programs:
+                - "decoder": Text generation decoder
+                - "token_embeddings": Token embedding layer
+                - "{modality}_encoder": Multimodal encoder (e.g., "audio_encoder")
+        """
+        with torch.no_grad():
+            # 1. Export text decoder.
+            exportable_module = TorchExportableModuleForDecoderOnlyLM(
+                self.model.language_model,
+            )
+            exported_programs = {}
+
+            # Custom SDPA for text decoder.
+            self._register_attention_mask(exportable_module)
+
+            if self.use_custom_kv_cache:
+                from optimum.executorch.attentions.custom_kv_cache import (
+                    replace_with_et_custom_kv_cache,
+                )
+
+                replace_with_et_custom_kv_cache(
+                    exportable_module.model,
+                    self.model.config.text_config,
+                    self.model.generation_config,
+                    self.model.dtype,
+                )
+
+            inputs_embeds, cache_position, dynamic_shapes = self._prepare_decoder_only_export_inputs()
+            logging.info(
+                f"Exporting decoder using inputs_embeds({inputs_embeds.shape}), cache_position({cache_position.shape})={cache_position}, dynamic_shapes={dynamic_shapes}"
+            )
+            exported_program = exportable_module.export(
+                inputs_embeds=inputs_embeds,
+                cache_position=cache_position,
+                dynamic_shapes=dynamic_shapes,
+                strict=True,
+            )
+            # Apply RemoveTransposes pass to remove
+            # any back-to-back transpose ops that are not needed
+            # e.g. output of update_cache is transposed and
+            # input to custom_sdpa is transposed.
+            from executorch.extension.llm.export.export_passes import (
+                RemoveRedundantTransposes,
+            )
+
+            mutated_gm = RemoveRedundantTransposes()(exported_program.module())[0]
+            exported_program = torch.export.export(
+                mutated_gm,
+                args=(),
+                kwargs={"cache_position": cache_position, "inputs_embeds": inputs_embeds},
+                dynamic_shapes=dynamic_shapes,
+                strict=True,
+            )
+            exported_programs["decoder"] = exported_program
+
+            # 2. Export token embeddings
+            input_ids, dynamic_shapes = self._prepare_text_embedding_export_inputs()
+            logging.info(
+                f"Exporting token embeddings using input_ids({input_ids.shape}), dynamic_shapes={dynamic_shapes}"
+            )
+
+            token_embeddings_exported_program = torch.export.export(
+                self.model.language_model.get_input_embeddings(),
```

Comment:
so language model export does not take care of embedding because we mix embedding for vision/audio part with language part?

Reply:
Yeah because it needs to be able to take in image embeddings too, so the token -> text embeddings part happens separately